docs/promtail-failure-modes.md

```diff
@@ -0,0 +1,57 @@
+# Promtail Failure Modes
+
+This document describes known failure modes of `promtail` on edge cases and the adopted trade-offs.
+
+
+## A tailed file is truncated while `promtail` is not running
+
+Given the following order of events:
+
+1. `promtail` is tailing `/app.log`
+2. `promtail` current position for `/app.log` is `100` (bytes)
+3. `promtail` is stopped
+4. `/app.log` is truncated and new logs are appended to it
+5. `promtail` is restarted
+
+When `promtail` is restarted, it reads the previous position (`100`) from the positions file. Two scenarios are then possible:
+
+- `/app.log` size is < than the position before truncating
+- `/app.log` size is >= than the position before truncating
+
+If the `/app.log` file size is less than the previous position, then the file is detected as truncated and logs will be tailed starting from position `0`. Otherwise, if the `/app.log` file size is >= than the previous position, `promtail` can't detect it was truncated while not running and will continue tailing the file from position `100`.
+
+
+## Loki ingester is unavailable
+
+For each tailing file, `promtail` reads a line, process it through the configured `pipeline_stages` and push the log entry to Loki ingester. Log entries are batched together before getting pushed to the ingester, based on the max batch duration `client.batch-wait` and size `client.batch-size-bytes`, whichever comes first.
+
+In case of any error while sending a log entries batch to the ingester, `promtail` adopts a "retry then discard" strategy:
+
+- `promtail` retries to send log entry to the ingester up to `maxretries` times
+- if all retries fail, `promtail` discards the batch of log entries (_which will be lost_) and proceeds with the next one
+
+You can configure the `maxretries` and the delay between two retries via the `backoff_config` in the promtail config file:
+
+```
+clients:
+  - url: INGESTER-URL
+    backoff_config:
+      minbackoff: 100ms
+      maxbackoff: 5s
+      maxretries: 5
+```
+
+
+## Duplicated logs after a `promtail` crash / panic
+
+When `promtail` shutdown gracefully, it saves the last read offsets in the positions file, so that on a subsequent restart it will continue tailing logs without duplicates neither losses.
+
+In the unlikely event of a crash, `promtail` can't save the last read offsets in the positions file. When restarted, `promtail` will read the positions file saved at the last sync period and will continue tailing the files from there. This means that if new log entries have been read and pushed to the ingester between the last sync period and the crash, these log entries will be sent again to the ingester on `promtail` restart.
```

Comment:
haha, i'm not sure how unlikely this is :) but I appreciate your optimism...  Though yes crashing is hopefully unlikely but OOM's in a kubernetes type environment could happen.  

Maybe also note that resending logs would be ignored by loki as loki currently rejects logs with older timestamps than it has already received?  So you don't need to crank down the sync_period, sending of some duplicates is ok

Reply:
> Maybe also note that resending logs would be ignored by loki as loki currently rejects logs with older timestamps than it has already received? So you don't need to crank down the sync_period, sending of some duplicates is ok

You're definitely right. I will revisit that paragraph accordingly.